File: 論文逐段精讀器/scripts/test_remaining_companions.py
#!/usr/bin/env python3
"""Regression checks for source retention, reconstructed units, and all seven crops."""
import copy
import hashlib
import json
import unittest
import fitz
from build_remaining_companions import ROOT, REVISION, core
from companion_metadata import reconcile_companion
from validate_phase_a import validate_paper
from audit_reading_quality import audit

class RemainingCompanionTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.papers={f'paper-{n:02d}':json.loads((ROOT/f'public/data/paper-{n:02d}.json').read_text()) for n in range(2,9)}
        cls.before={pid:json.loads((ROOT/f'content/{pid}-before-remaining-companion.json').read_text()) for pid in cls.papers}

    def test_sources_ids_and_reference_material_retained(self):
        for pid,p in self.papers.items():
            with self.subTest(paper=pid):
                old=self.before[pid];current={s['id']:s for s in p['segments']}
                baseline=ROOT/f'content/{pid}-before-remaining-companion.json'
                self.assertEqual(hashlib.sha256(baseline.read_bytes()).hexdigest(),p['remainingPaperProcessing']['beforeDataSha256'])
                self.assertEqual(p['sourceSha256'],old['sourceSha256']);self.assertEqual(p['references'],old['references'])
                self.assertEqual(p['citations'][:len(old['citations'])],old['citations'])
                provenance={s['id']:s for r in p['readingUnitRepairs'] if r['revision']==REVISION for s in r['previousSegments']}
                for s in old['segments']:
                    self.assertIn(s['id'],current)
                    if current[s['id']]['sourceText']!=s['sourceText'] or current[s['id']]['fragments']!=s['fragments']:
                        self.assertEqual(provenance[s['id']]['sourceText'],s['sourceText'])
                        self.assertEqual(provenance[s['id']]['fragments'],s['fragments'])
                self.assertEqual(p['publicationStatus'],'draft')

    def test_every_dataset_and_companion_are_valid(self):
        total=0;library=json.loads((ROOT/'public/data/manifest.json').read_text())
        for pid,p in self.papers.items():
            with self.subTest(paper=pid):
                validate_paper(p,ROOT/'public')
                self.assertEqual(p['exhibitCompanion']['revision'],REVISION)
                self.assertEqual(len(p['exhibitCompanion']['studies']),len(p['exhibits']))
                self.assertEqual(audit(p)['errors'],0)
                self.assertFalse(any(f['category']=='table-cell-as-body' for f in audit(p)['findings']))
                self.assertTrue(all(s['translation']['faithfulZh'].strip() and s['translation']['plainZh'].strip() for s in core(p)))
                self.assertTrue(all('```' not in s['translation']['plainZh'] and '"explanation":' not in s['translation']['plainZh'] for s in core(p)))
                entry=next(e for e in library['papers'] if e['id']==pid)
                self.assertEqual(entry['bodySegmentCount'],len(core(p)));self.assertEqual(entry['segmentCount'],len(p['segments']))
                total+=len(p['exhibits'])
        self.assertEqual(total,107)

    def test_complete_300dpi_images(self):
        for pid,p in self.papers.items():
            with fitz.open(ROOT/'public'/p['pdfUrl'].lstrip('/')) as pdf:
                for e in p['exhibits']:
                    with self.subTest(exhibit=e['id']):
                        pix=fitz.Pixmap(ROOT/'public'/e['imageUrl'].lstrip('/'))
                        self.assertAlmostEqual(pix.width,(e['bbox'][2]-e['bbox'][0])*300/72,delta=2)
                        self.assertAlmostEqual(pix.height,(e['bbox'][3]-e['bbox'][1])*300/72,delta=2)
                        self.assertGreater(pix.width,500);self.assertGreater(pix.height,100)
                        if e['kind']=='figure':
                            images=pdf[e['page']-1].get_image_info()
                            self.assertTrue(any((fitz.Rect(e['bbox'])&fitz.Rect(i['bbox'])).get_area()/fitz.Rect(i['bbox']).get_area()>.99 for i in images))

    def test_old_unedited_drafts_migrate_without_discarding_records(self):
        for pid,p in self.papers.items():
            with self.subTest(paper=pid):
                migrated=reconcile_companion(self.before[pid],p)
                validate_paper(migrated,ROOT/'public')
                self.assertEqual({s['id'] for s in migrated['segments']},{s['id'] for s in p['segments']})
                self.assertEqual([s['id'] for s in core(migrated)],[s['id'] for s in core(p)])
                repair=next(r for r in p['readingUnitRepairs'] if r['revision']==REVISION and r['absorbedIds'])
                edited=copy.deepcopy(self.before[pid]);old=next(s for s in edited['segments'] if s['id']==repair['absorbedIds'][0]);old['translation']['plainZh']='保留我的手動稿'
                result=reconcile_companion(edited,p)
                self.assertIn(repair['targetId'],result['pendingReadingUnitRepairs'])
                self.assertEqual(next(s for s in result['segments'] if s['id']==old['id'])['translation']['plainZh'],'保留我的手動稿')

    def test_known_source_recoveries_and_discrepancies_are_visible(self):
        p3=self.papers['paper-03'];self.assertTrue(any(s['sourceText'].startswith('In physical education (PE)') for s in core(p3)))
        p8=self.papers['paper-08'];self.assertEqual(sum(e['kind']=='figure' for e in p8['exhibits']),11)
        self.assertTrue(any(s['sourceRecovery']['method'].startswith('Tesseract') for s in core(p8) if s.get('sourceRecovery')))
        for number,statistic in [('2','t = 2.30'),('3','t = 2.22'),('4','t = 2.11'),('5','t = 2.40')]:
            self.assertIn(statistic,p8['exhibitCompanion']['studies']['paper-08-table-'+number]['summaryZh'])
        self.assertTrue(any('9 次' in text and '7 次' in text for text in p8['exhibitCompanion']['studies']['paper-08-figure-10']['limitsZh']))
        p6=self.papers['paper-06'];self.assertEqual(sum(l['method']=='semantic' and l['confidence']=='high' for l in p6['exhibitCompanion']['links']),3)
        p5=self.papers['paper-05'];self.assertEqual(len(p5['readingQuality']['appendices']),1)
        for label in('A1–08','A2–10','A3–11','A4–07'):self.assertIn(label,p5['readingQuality']['appendices'][0]['faithfulZh'])

if __name__=='__main__':unittest.main()

File: 論文逐段精讀器/scripts/test_remaining_companions.mjs
#!/usr/bin/env node
import assert from 'node:assert/strict';
import {readFile} from 'node:fs/promises';
import ts from 'typescript';
const read=async name=>JSON.parse(await readFile(new URL(name,import.meta.url),'utf8'));
const compile=async name=>'data:text/javascript;base64,'+Buffer.from(ts.transpileModule(await readFile(new URL(name,import.meta.url),'utf8'),{compilerOptions:{module:ts.ModuleKind.ESNext,target:ts.ScriptTarget.ES2022}}).outputText).toString('base64');
const {loadCuratorDraft,curatorStorageKey}=await import(await compile('../lib/curator-state.ts'));
const {readingSegments,resolveReadingSegment}=await import(await compile('../lib/companion-state.ts'));
const {explanationIssues}=await import(await compile('../lib/reading-quality.ts'));
const storage=new Map();globalThis.window={localStorage:{getItem:key=>storage.get(key)??null,setItem:(key,value)=>storage.set(key,value)}};
for(let n=2;n<=8;n++){
 const pid=`paper-${String(n).padStart(2,'0')}`,paper=await read(`../public/data/${pid}.json`),before=await read(`../content/${pid}-before-remaining-companion.json`);
 storage.set(curatorStorageKey(pid),JSON.stringify(before));const migrated=loadCuratorDraft(paper);
 assert.deepEqual(readingSegments(migrated).map(s=>s.id),readingSegments(paper).map(s=>s.id),pid);
 assert.deepEqual(migrated.segments.map(s=>s.id),paper.segments.map(s=>s.id),pid);
 for(const r of paper.readingUnitRepairs.filter(r=>r.revision==='remaining-papers-companion-v1')){
  const target=migrated.segments.find(s=>s.id===r.targetId);assert.equal(target.sourceText,paper.segments.find(s=>s.id===r.targetId).sourceText);
  for(const id of r.absorbedIds){const source=migrated.segments.find(s=>s.id===r.targetId);if(!source.excluded&&!source.readingRole)assert.equal(resolveReadingSegment(migrated,id)?.id,r.targetId);}
 }
 const repair=paper.readingUnitRepairs.find(r=>r.revision==='remaining-papers-companion-v1'&&r.absorbedIds.length);
 const edited=structuredClone(before);edited.segments.find(s=>s.id===repair.absorbedIds[0]).translation.plainZh='保留我的手動稿';storage.set(curatorStorageKey(pid),JSON.stringify(edited));
 const preserved=loadCuratorDraft(paper);assert.ok(preserved.pendingReadingUnitRepairs.includes(repair.targetId));assert.equal(preserved.segments.find(s=>s.id===repair.absorbedIds[0]).translation.plainZh,'保留我的手動稿');
 for(const s of readingSegments(paper))assert.deepEqual(explanationIssues(paper,s),[],s.id);
}
console.log('Seven papers: safe curator migrations, archived resume points, protected edits and glossary checks passed.');
